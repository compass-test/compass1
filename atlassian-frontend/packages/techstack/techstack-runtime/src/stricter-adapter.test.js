const isFunction = require('lodash/isFunction');
const mapValues = require('lodash/mapValues');

const stricterAdapter = require('./stricter-adapter');

const createConfiguration = ({ pluginId, configId }) => ({
  level: 'error',
  config: { foo: `bar-${pluginId}-${configId}` },
});

const createCheck = ({ pluginId, configId, noPlugin = false }) => ({
  checks: [
    {
      plugin: !noPlugin ? `plugin-${pluginId}` : undefined,
      rule: `rule-${pluginId}`,
      configuration: createConfiguration({ pluginId, configId }),
    },
  ],
  packageRootPath: {
    relative: `foo`,
    absolute: `/Foo/project/packages/foo`,
  },
});

const mockChecks = [
  createCheck({ pluginId: 'a', configId: 'a' }),
  createCheck({ pluginId: 'a', configId: 'b' }),
  createCheck({ pluginId: 'b', configId: 'a' }),
  createCheck({ pluginId: 'c', configId: 'a' }),
  createCheck({ pluginId: 'c', configId: 'b' }),
  createCheck({ pluginId: 'd', configId: 'a', noPlugin: true }),
];

jest.mock('./utils', () => ({
  getChecksForPackages: () => mockChecks,
  getTechstackConfig: () => ({
    config: {
      pathToPackages: 'packages',
      rootPath: '.',
    },
  }),
}));

describe('Stricter adapter', () => {
  test('returns the correct stricter config', () => {
    expect(
      stricterAdapter({
        plugins: ['plugin-e'],
        rules: {
          'rule-e': createConfiguration({ pluginId: 'e', configId: 'a' }),
        },
      }),
    ).toMatchSnapshot();
  });

  test('returns the correct stricter config when existing rules clash with the techstack rules', () => {
    const result = stricterAdapter({
      plugins: ['plugin-a', 'plugin-b', 'plugin-c'],
      rules: {
        'rule-a': createConfiguration({ pluginId: 'a', configId: 'c' }),
        'rule-b': [
          createConfiguration({ pluginId: 'b', configId: 'b' }),
          createConfiguration({ pluginId: 'b', configId: 'c' }),
        ],
        'rule-c': () => createConfiguration({ pluginId: 'c', configId: 'c' }),
        'rule-d': () => [
          createConfiguration({ pluginId: 'd', configId: 'b' }),
          createConfiguration({ pluginId: 'd', configId: 'c' }),
        ],
      },
    });

    expect(result).toMatchSnapshot();

    const resolvedRules = mapValues(result.rules, val =>
      isFunction(val) ? val({ packages: [] }) : val,
    );

    expect(resolvedRules).toMatchSnapshot();
  });
});
