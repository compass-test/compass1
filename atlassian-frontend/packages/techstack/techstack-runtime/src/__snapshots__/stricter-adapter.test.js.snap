// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stricter adapter returns the correct stricter config when existing rules clash with the techstack rules 1`] = `
Object {
  "plugins": Array [
    "plugin-a",
    "plugin-b",
    "plugin-c",
  ],
  "rules": Object {
    "rule-a": Array [
      Object {
        "config": Object {
          "foo": "bar-a-a",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
      Object {
        "config": Object {
          "foo": "bar-a-b",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
      Object {
        "config": Object {
          "foo": "bar-a-c",
        },
        "level": "error",
      },
    ],
    "rule-b": Array [
      Object {
        "config": Object {
          "foo": "bar-b-a",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
      Object {
        "config": Object {
          "foo": "bar-b-b",
        },
        "level": "error",
      },
      Object {
        "config": Object {
          "foo": "bar-b-c",
        },
        "level": "error",
      },
    ],
    "rule-c": [Function],
    "rule-d": [Function],
  },
}
`;

exports[`Stricter adapter returns the correct stricter config when existing rules clash with the techstack rules 2`] = `
Object {
  "rule-a": Array [
    Object {
      "config": Object {
        "foo": "bar-a-a",
      },
      "include": Array [
        /\\^foo\\\\//,
      ],
      "level": "error",
    },
    Object {
      "config": Object {
        "foo": "bar-a-b",
      },
      "include": Array [
        /\\^foo\\\\//,
      ],
      "level": "error",
    },
    Object {
      "config": Object {
        "foo": "bar-a-c",
      },
      "level": "error",
    },
  ],
  "rule-b": Array [
    Object {
      "config": Object {
        "foo": "bar-b-a",
      },
      "include": Array [
        /\\^foo\\\\//,
      ],
      "level": "error",
    },
    Object {
      "config": Object {
        "foo": "bar-b-b",
      },
      "level": "error",
    },
    Object {
      "config": Object {
        "foo": "bar-b-c",
      },
      "level": "error",
    },
  ],
  "rule-c": Array [
    Object {
      "config": Object {
        "foo": "bar-c-a",
      },
      "include": Array [
        /\\^foo\\\\//,
      ],
      "level": "error",
    },
    Object {
      "config": Object {
        "foo": "bar-c-b",
      },
      "include": Array [
        /\\^foo\\\\//,
      ],
      "level": "error",
    },
    Object {
      "config": Object {
        "foo": "bar-c-c",
      },
      "level": "error",
    },
  ],
  "rule-d": Array [
    Object {
      "config": Object {
        "foo": "bar-d-a",
      },
      "include": Array [
        /\\^foo\\\\//,
      ],
      "level": "error",
    },
    Object {
      "config": Object {
        "foo": "bar-d-b",
      },
      "level": "error",
    },
    Object {
      "config": Object {
        "foo": "bar-d-c",
      },
      "level": "error",
    },
  ],
}
`;

exports[`Stricter adapter returns the correct stricter config 1`] = `
Object {
  "plugins": Array [
    "plugin-e",
    "plugin-a",
    "plugin-b",
    "plugin-c",
  ],
  "rules": Object {
    "rule-a": Array [
      Object {
        "config": Object {
          "foo": "bar-a-a",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
      Object {
        "config": Object {
          "foo": "bar-a-b",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
    ],
    "rule-b": Array [
      Object {
        "config": Object {
          "foo": "bar-b-a",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
    ],
    "rule-c": Array [
      Object {
        "config": Object {
          "foo": "bar-c-a",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
      Object {
        "config": Object {
          "foo": "bar-c-b",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
    ],
    "rule-d": Array [
      Object {
        "config": Object {
          "foo": "bar-d-a",
        },
        "include": Array [
          /\\^foo\\\\//,
        ],
        "level": "error",
      },
    ],
    "rule-e": Object {
      "config": Object {
        "foo": "bar-e-a",
      },
      "level": "error",
    },
  },
}
`;
